File: projects/advent_of_code/solutions/year2025/day02/go/solution.go
package main

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/jomcgi/homelab/projects/advent_of_code/pkg/aoc"
)

func main() {
	input, err := aoc.ReadFile("projects/advent_of_code/solutions/year2025/day02/input.txt")
	if err != nil {
		panic(err)
	}

	fmt.Println("Part 1:", part1(input))
	fmt.Println("Part 2:", part2(input))
}

func part1(input string) int {
	// Parse all ranges from input (comma-separated, each is start-end)
	ranges := parseRanges(input)

	sum := 0
	for _, r := range ranges {
		sum += sumInvalidInRange(r[0], r[1])
	}
	return sum
}

func parseRanges(input string) [][2]int {
	input = strings.TrimSpace(input)
	parts := strings.Split(input, ",")
	var ranges [][2]int
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p == "" {
			continue
		}
		nums := aoc.Ints(p)
		if len(nums) == 2 {
			ranges = append(ranges, [2]int{nums[0], nums[1]})
		}
	}
	return ranges
}

// isInvalidID checks if a number is a "repeated twice" pattern
// e.g., 55 (5+5), 6464 (64+64), 123123 (123+123)
func isInvalidID(n int) bool {
	s := strconv.Itoa(n)
	if len(s)%2 != 0 {
		return false
	}
	half := len(s) / 2
	return s[:half] == s[half:]
}

// sumInvalidInRange finds all invalid IDs in [start, end] and returns their sum
func sumInvalidInRange(start, end int) int {
	sum := 0
	// Generate all possible "doubled" numbers and check if they're in range
	// This is more efficient than checking every number in the range
	for digits := 1; digits <= 10; digits++ {
		// Generate all doubled numbers with `digits` in each half
		minBase := 1
		if digits > 1 {
			minBase = pow10(digits - 1)
		}
		maxBase := pow10(digits) - 1

		for base := minBase; base <= maxBase; base++ {
			// Create doubled number: base repeated twice
			doubled := base*pow10(digits) + base
			if doubled >= start && doubled <= end {
				sum += doubled
			}
			if doubled > end {
				break
			}
		}
	}
	return sum
}

func pow10(n int) int {
	result := 1
	for i := 0; i < n; i++ {
		result *= 10
	}
	return result
}

func part2(input string) int {
	ranges := parseRanges(input)

	sum := 0
	for _, r := range ranges {
		sum += sumInvalidInRangeV2(r[0], r[1])
	}
	return sum
}

// isInvalidIDV2 checks if a number is a pattern repeated at least twice
// e.g., 55 (5×2), 123123 (123×2), 123123123 (123×3), 1111111 (1×7)
func isInvalidIDV2(n int) bool {
	s := strconv.Itoa(n)
	length := len(s)

	// Try all possible base lengths that divide the total length
	// and result in at least 2 repetitions
	for baseLen := 1; baseLen <= length/2; baseLen++ {
		if length%baseLen != 0 {
			continue
		}
		reps := length / baseLen
		if reps < 2 {
			continue
		}

		base := s[:baseLen]
		valid := true
		for i := 1; i < reps; i++ {
			if s[i*baseLen:(i+1)*baseLen] != base {
				valid = false
				break
			}
		}
		if valid {
			return true
		}
	}
	return false
}

// sumInvalidInRangeV2 finds all invalid IDs (repeated at least twice) in [start, end]
func sumInvalidInRangeV2(start, end int) int {
	seen := make(map[int]bool)

	// Calculate the multiplier for a given base length and reps
	// e.g., for baseDigits=2, reps=3: multiplier = 1 + 100 + 10000 = 10101
	getMultiplier := func(baseDigits, reps int) int {
		mult := 0
		p := 1
		for r := 0; r < reps; r++ {
			mult += p
			p *= pow10(baseDigits)
		}
		return mult
	}

	// For each base length
	for baseDigits := 1; baseDigits <= 10; baseDigits++ {
		minBase := 1
		if baseDigits > 1 {
			minBase = pow10(baseDigits - 1)
		}
		maxBase := pow10(baseDigits) - 1

		// Try different repetition counts (at least 2)
		for reps := 2; reps <= 20; reps++ {
			multiplier := getMultiplier(baseDigits, reps)

			// Check if any number in this category could be in range
			minNum := minBase * multiplier
			maxNum := maxBase * multiplier

			if minNum > end {
				break // All larger reps will also be too big
			}
			if maxNum < start {
				continue // This category is entirely below range
			}

			// Calculate the range of bases that produce numbers in [start, end]
			loBase := (start + multiplier - 1) / multiplier // ceiling division
			if loBase < minBase {
				loBase = minBase
			}
			hiBase := end / multiplier
			if hiBase > maxBase {
				hiBase = maxBase
			}

			for base := loBase; base <= hiBase; base++ {
				num := base * multiplier
				if num >= start && num <= end && !seen[num] {
					seen[num] = true
				}
			}
		}
	}

	sum := 0
	for num := range seen {
		sum += num
	}
	return sum
}

File: projects/advent_of_code/solutions/year2025/day02/python/solution.py
#!/usr/bin/env python3
"""Advent of Code 2025 Day 2"""

from projects.advent_of_code.python.aoc import read_file


def parse_ranges(input_text: str) -> list[tuple[int, int]]:
    """Parse comma-separated ranges from input."""
    ranges = []
    for part in input_text.strip().split(","):
        part = part.strip()
        if part and "-" in part:
            # Split on dash to get start-end (can't use ints() as it treats - as negative)
            start_str, end_str = part.split("-", 1)
            ranges.append((int(start_str), int(end_str)))
    return ranges


def sum_invalid_in_range(start: int, end: int) -> int:
    """Find all 'doubled' numbers (base repeated twice) in [start, end]."""
    total = 0
    for digits in range(1, 11):
        min_base = 1 if digits == 1 else 10 ** (digits - 1)
        max_base = 10**digits - 1
        multiplier = 10**digits + 1  # e.g., for 2 digits: 101

        for base in range(min_base, max_base + 1):
            num = base * multiplier
            if num > end:
                break
            if num >= start:
                total += num
    return total


def sum_invalid_in_range_v2(start: int, end: int) -> int:
    """Find all numbers where a base is repeated 2+ times in [start, end]."""
    seen = set()

    for base_digits in range(1, 11):
        min_base = 1 if base_digits == 1 else 10 ** (base_digits - 1)
        max_base = 10**base_digits - 1

        for reps in range(2, 21):
            # Calculate multiplier: 1 + 10^d + 10^(2d) + ... for reps terms
            multiplier = sum(10 ** (base_digits * r) for r in range(reps))

            min_num = min_base * multiplier
            max_num = max_base * multiplier

            if min_num > end:
                break
            if max_num < start:
                continue

            # Calculate range of bases that produce numbers in [start, end]
            lo_base = max(min_base, (start + multiplier - 1) // multiplier)
            hi_base = min(max_base, end // multiplier)

            for base in range(lo_base, hi_base + 1):
                num = base * multiplier
                if start <= num <= end:
                    seen.add(num)

    return sum(seen)


def part1(input_text: str) -> int:
    ranges = parse_ranges(input_text)
    return sum(sum_invalid_in_range(start, end) for start, end in ranges)


def part2(input_text: str) -> int:
    ranges = parse_ranges(input_text)
    return sum(sum_invalid_in_range_v2(start, end) for start, end in ranges)


if __name__ == "__main__":
    input_text = read_file("advent_of_code/solutions/year2025/day02/input.txt")

    print(f"Part 1: {part1(input_text)}")
    print(f"Part 2: {part2(input_text)}")
